src/event/finalization.rs

```diff
@@ -0,0 +1,358 @@
+        }
+    }
+
+    /// Update all sources for this finalizer with the current
+    /// status. This *drops* the finalizer array elements so they may
+    /// immediately signal the source batch.
+    pub fn update_sources(&mut self) {
+        let finalizers = mem::replace(&mut self.0, vec![].into());
+        for finalizer in finalizers.iter() {
+            finalizer.update_batch();
+        }
+    }
+
+    #[cfg(test)]
+    fn count_finalizers(&self) -> usize {
+        self.0.len()
+    }
+}
+
+/// An event finalizer is the shared data required to handle tracking
+/// the status of an event, and updating the status of a batch with that
+/// when the event is dropped.
+#[derive(Debug)]
+pub struct EventFinalizer {
+    status: Atomic<EventStatus>,
+    batch: Arc<BatchNotifier>,
+}
+
+impl EventFinalizer {
+    /// Create a new event in a batch.
+    pub fn new(batch: Arc<BatchNotifier>) -> Self {
+        let status = Atomic::new(EventStatus::Dropped);
+        Self { status, batch }
+    }
+
+    /// Update this finalizer's status in place with the given `EventStatus`
+    pub fn update_status(&self, status: EventStatus) {
+        self.status
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |old_status| {
+                Some(old_status.update(status))
+            })
+            .unwrap_or_else(|_| unreachable!());
+    }
+
+    /// Update the batch for this event with this finalizer's
+    /// status, and mark this event as no longer requiring update.
+    pub fn update_batch(&self) {
+        let status = self
+            .status
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |_| {
+                Some(EventStatus::NoOp)
+            })
+            .unwrap_or_else(|_| unreachable!());
+        self.batch.update_status(status);
+    }
+}
+
+impl Drop for EventFinalizer {
+    fn drop(&mut self) {
+        self.update_batch();
+    }
+}
+
+/// A batch notifier contains the status of the current batch along with
+/// a one-shot notifier to send that status back to the source. It is
+/// shared among all events of a batch.
+#[derive(Debug)]
+pub struct BatchNotifier {
+    status: Atomic<BatchStatus>,
+    notifier: Option<oneshot::Sender<BatchStatus>>,
+}
+
+impl BatchNotifier {
+    /// Create a new `BatchNotifier` along with the receiver used to
+    /// await its finalization status.
+    pub fn new_with_receiver() -> (Arc<Self>, oneshot::Receiver<BatchStatus>) {
+        let (sender, receiver) = oneshot::channel();
+        let notifier = Self {
+            status: Atomic::new(BatchStatus::Delivered),
+            notifier: Some(sender),
+        };
+        (Arc::new(notifier), receiver)
+    }
+
+    /// Update this notifier's status from the status of a finalized event.
+    pub fn update_status(&self, status: EventStatus) {
+        self.status
+            .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |old_status| {
+                Some(old_status.update_from_event(status))
```

Comment:
`update_from_event` could be inlined here which would give us something like
```
if status==EventStatus::Failed {
    self.status.store(Ordering::Relaxed, BatchStatus::Failed);
}
```

Reply:
This makes sense as an optimization, I had hoped to keep the logic of how the `BatchStatus` may be changed scoped to its implementation block. Adding a newtype wrapper for `Atomic<BatchStatus>` is probably not worth it. I'll add this.